synIndex.R

```diff
@@ -25,44 +34,119 @@ source('params.R')
 # we will work with that
 
-## Get a list of all local files
-localFileList <- list.files(path = AWS_DOWNLOAD_LOCATION,
-                            all.files = TRUE, # get hidden files too
-                            recursive = TRUE, # get files inside sub-folders (if any)
-                            full.names = FALSE) # to get the directory path prepended to the file name
+# all folders inside the AWS_DOWNLOAD_LOCATION
+# When we create dataFileHandleId in synapse it will create some folders in the source S3 bucket
+# we don't need those
+localDirs <- list.dirs(path = AWS_DOWNLOAD_LOCATION, full.names = FALSE) 
+
+## The folders we want to show in synapse
+## This would depend on the how the data is organized in the main S3 INGRESS bucket
+FOLDERS_TO_SYNC_SYNAPSE <- c('adults',
+                             'pregnant',
+                             'pediatric') 
+
+## From the local AWS location remove all folders that are not the ones selected above
+dirs_to_delete <- dplyr::setdiff(localDirs, c(FOLDERS_TO_SYNC_SYNAPSE, "")) # "" is the local directory, a result of list.dirs(.. full.names=FALSE)
+
+## delete the unwanted directories before creating a upload manifest
+for(dir_ in dirs_to_delete){
+  unlink(paste0(AWS_DOWNLOAD_LOCATION,"/",dir_), recursive = TRUE)
+}
+
+# localFileList <- list.files(path = AWS_DOWNLOAD_LOCATION,
+#                             all.files = TRUE, # get hidden files too
+#                             recursive = TRUE, # get files inside sub-folders (if any)
+#                             full.names = FALSE) # to get the directory path prepended to the file name
+
+#############
+# Get a manifest of files to upload
+#############
+## The S3 bucket is synced to AWS_DOWNLOAD_LOCATION locally. 
+## We will use synapse cmd line client manifest function to replicate the folder structure 
+## but NOT upload the files. We will create a datafilehandleid in place later instead of uploading file
+
+old_path <- Sys.getenv("PATH")
+Sys.setenv(PATH = paste(old_path, "/home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/PythonEmbedInR/bin", sep = ":"))
+## When we install synapser, it installs synapseclient and along with it the synapse cmd line client
+## We need to add the location of synapseclient to the system path so that it can recognize synapse cmd 
+
+SYSTEM_COMMAND <- glue::glue('synapse -u "{SYNAPSE_USERNAME}" -p "{SYNAPSE_PASSWORD}" manifest --parent-id {SYNAPSE_PARENT_ID} --manifest current_manifest.tsv {AWS_DOWNLOAD_LOCATION}')
+
+## Generate manifest file
+system(SYSTEM_COMMAND)
+
+###########
+## Get a list of all files to upload and their synapse locations(parentId) 
+###########
+STR_LEN_AWS_DOWNLOAD_LOCATION = stringr::str_length(AWS_DOWNLOAD_LOCATION)
+
+## All files present locally from manifest
+synapse_manifest <- read.csv('current_manifest.tsv', sep = '\t', stringsAsFactors = F) %>% 
+  dplyr::filter(path != paste0(AWS_DOWNLOAD_LOCATION,'/owner.txt')) %>%  # need not create a dataFileHandleId for owner.txt
```

Comment:
What happens if you have run this once and the file current_manifest.tsv was generated then you run again but the system command didn't work properly. Will the script use the out of date manifest file. 

I think that the manifest file generated should be a temporary file that is unique for every execution.

Reply:
An old manifest will update no files in synapse if there are no new files. I checked this.